sklearn/linear_model/_cd_fast.pyx

```diff
@@ -264,9 +264,9 @@ def sparse_enet_coordinate_descent(
     floating [::1] w,
     floating alpha,
     floating beta,
-    np.ndarray[floating, ndim=1, mode='c'] X_data,
-    np.ndarray[int, ndim=1, mode='c'] X_indices,
-    np.ndarray[int, ndim=1, mode='c'] X_indptr,
+    floating[::1] X_data,
```

Comment:
Not sure it works with readonly X. Can you test this against a read-only csr matrix (``Xcsr.data.setflags(write=False)``) ?

Reply:
Can confirm it works with read-only CSR matrices:
```python
from scipy import sparse
from sklearn import linear_model
import numpy as np

clf = linear_model.ElasticNet(alpha=0.1)
X = sparse.random(100,10)
X.data.setflags(write=False)
y = np.random.random(100)
clf.fit(X, y)
```

Reply:
By default ElasticNet does a copy of the input. It will also convert if it's not in csc format. If I change your snippet to
```python
from scipy import sparse
from sklearn import linear_model
import numpy as np

clf = linear_model.ElasticNet(alpha=0.1, copy_X=False)
X = sparse.random(100,10, format="csc")
X.data.setflags(write=False)
y = np.random.random(100)
clf.fit(X, y)
```
Then I get ``ValueError: buffer source array is read-only``

Reply:
Ah gotcha. In that case, since `sparse_enet_coordinate_descent` doesn't actually modify `X_data`, I went ahead and implemented the fix suggested in #10624, which is the `ReadonlyArrayWrapper` added in #20903. I think this should work, but I'm still quite new to the Cython side of things, so please let me know if there are problems with this approach -- thanks :)

Reply:
Looks good. I guess we can do the same for ``enet_coordinate_descent`` because currently we have the same error for the dense case:
```python
from scipy import sparse
from sklearn import linear_model
import numpy as np

clf = linear_model.ElasticNet(alpha=0.1, copy_X=False)
X = np.asfortranarray(np.random.uniform(size=(100, 10)))
X.setflags(write=False)
y = np.random.random(100)
clf.fit(X, y)
```

and probably also ``enet_coordinate_descent_multi_task``. It would be good to also add a test for these read-only cases because it doesn't seem tested currently

Reply:
Can `indices` or `indptr` be read only?

I think testing is a must here. If we pass actual read only data with `ReadOnlyArrayWrapper`, python crashes if the Cython function writes to the memoryview:

```python
%%cython
from cython cimport floating

def f(floating[:] a):
    a[0] = 1
``` 

```python
from sklearn.utils._readonly_array_wrapper import ReadonlyArrayWrapper
from sklearn.utils._testing import create_memmap_backed_data
import numpy as np

X = np.asarray([1, 2.3])
X_mapped = create_memmap_backed_data(X)

# crashes
f(ReadonlyArrayWrapper(X_mapped))
```

Reply:
Added a test for this behavior and confirmed the algorithm doesn't modify `indices` or `indptr` so extended the wrapper to them as well.